source/common/common/token_bucket_impl.cc

```diff
@@ -11,10 +11,19 @@ TokenBucketImpl::TokenBucketImpl(uint64_t max_tokens, TimeSource& time_source, d
 uint64_t TokenBucketImpl::consume(uint64_t tokens, bool allow_partial) {
   if (tokens_ < max_tokens_) {
     const auto time_now = time_source_.monotonicTime();
-    tokens_ = std::min((std::chrono::duration<double>(time_now - last_fill_).count() * fill_rate_) +
-                           tokens_,
-                       max_tokens_);
-    last_fill_ = time_now;
+
+    const uint64_t new_fill_tokens =
+        std::chrono::duration_cast<std::chrono::microseconds>(time_now - last_fill_).count() *
```

Comment:
I feel that the main change here is less coarse updates to `tokens_`. Couldn't we keep the implementation as is but increase the resolution to microseconds (`std::chrono::duration<double, std::micro>(time_now - last_fill_)`)?

Reply:
Oops, we'd need to account for the change in scale of the number of available tokens. Could be work around floating point multiplication/division by scaling up the number of requested tokens instead.

Reply:
Would use of names like new_fill_micro_tokens and residual_micro_tokens_ help with understanding of the unit conversions involved in this computation?

Reply:
I don't think the changes are hard to understand actually; I think the original implementation is clearer though and only needs minor changes to reduce its coarseness.

Reply:
> Would use of names like new_fill_micro_tokens and residual_micro_tokens_ help with understanding of the unit conversions involved in this computation?

Yes  I think this name is well.

Reply:
> > Would use of names like new_fill_micro_tokens and residual_micro_tokens_ help with understanding of the unit conversions involved in this computation?
> 
> Yes I think this name is well.



> I don't think the changes are hard to understand actually; I think the original implementation is clearer though and only needs minor changes to reduce its coarseness.

Yes, the original implementation is too rough and does not take into account the resident token, which is why I mentioned this PR。

Reply:
> I feel that the main change here is less coarse updates to `tokens_`. Couldn't we keep the implementation as is but increase the resolution to microseconds (`std::chrono::duration<double, std::micro>(time_now - last_fill_)`)? Less code -- easier to follow.
> 
> We don't do any division/multiplication on doubles in the original implementation, performance-wise it shouldn't be far off the proposed change.

I think std::chrono::duration_cast<micro...> is better than std::chrono::duration

fisrst, std::chrono::duration_cast<micro...> Clearer semantics
second， std::chrono::duration_cast theoretically more efficient

Within gcc8.3, std::chrono::duration implement as follow
`
template<typename _Rep, typename _Period = ratio<1>>
      struct duration;
// Your implement call this
template<typename _Rep2, typename _Period2, typename = _Require<
		 __or_<__is_float<rep>,
		       __and_<__is_harmonic<_Period2>,
			      __not_<__is_float<_Rep2>>>>>>
	  constexpr duration(const duration<_Rep2, _Period2>& __d)
	  : __r(duration_cast<duration>(__d).count()) { } // this call duration_cast
constexpr rep count() const { 
     return __r; 
}
`